content/0000-puppet-integration/rfc.md

```diff
@@ -0,0 +1,420 @@
+* **Performers:** Noah Fontes, Hunter Haugen
+* **Inputers:** Rick Lane, Kyle Terry
+* **Deciders:** Hunter Haugen, Deepak Giridharagopal
+
+## Problem
+
+Puppet and Relay are currently two distinct product offerings with little
+interoperability. This is not optimal for our portfolio, and we hope to address
+the potential symbiotic relationship between the two platforms with an initial
+alpha-quality integration strategy.
+
+## Summary
+
+This RFC defines a structure for a Puppet module, an auxiliary process that acts
+as a sidecar to the Puppet server, and a new connection type in Relay. These
+three components provide the basis for bidirectional communication between an
+open-source Puppet or Puppet Enterprise (PE) install and Relay.
+
+## Motivation
+
+Relay provides an opportunity to bolster existing Puppet solutions with an
+event-based execution model that could make operating systems supervised by
+Puppet substanitally easier. We want to investigate what those opportunities
+look like to make the best broad-spectrum portfolio decisions possible.
+
+## Product-level explanation
+
+This RFC provides an implementation plan for initially integrating with
+open-source Puppet and PE. We intend to demonstrate the technical feasibility of
+an integration as well as demonstrating clear value for existing Puppet users in
+adopting Relay.
+
+We will initially provide these Relay workflows to integrate with Puppet:
+
+1. Notification to a Slack channel when a Puppet run detects configuration
+   changes for a given resource or profile
+1. Dry run of Puppet on a subset of nodes, manual approval, and dispatching a
+   real run back to Puppet
+
+Puppet users will be required to install a Puppet module,
+[puppetlabs-relay](https://github.com/puppetlabs/puppetlabs-relay), which will
+provide on-premise connectivity to the Relay environment. Within Relay, users
+will need to set up workflows with push triggers to receive data from Puppet and
+a special connection type to send data to Puppet.
+
+## Engineering-level explanation
+
+To provide bi-directional communication between on-premise Puppet installs and
+the Relay SaaS service, we'll use a Puppet module that includes plugins to send
+data to workflows via triggers (outbound) and a scheduled process that polls for
+requests to run Puppet on specific nodes (inbound). In particular, we want to
+avoid complex or undesirable networking scenarios like maintaining an
+out-of-band VPN connection or opening ports on corporate firewalls.
+
+### Module structure
+
+Our Puppet module defines one plugin, a report processor to run Relay workflows,
+and one class, `relay::bridge`, that installs the report processor (using
+[`ini_setting`](https://github.com/puppetlabs/puppetlabs-inifile) or
+`pe_ini_setting`) and a bidirectional process to handle requests to run Puppet.
+
+### Report processor
+
+The report processor sends report data to one or more workflow push triggers.
```

Comment:
Currently the push triggers are 1:1 token:workflow, and the report processor just takes a single token. I can update this to take a list of tokens, or just wait until the puppet connection is ready.